server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -527,106 +527,114 @@ public ParsedDocument index(IndexItem item) throws IOException {
+
+            for (var indexColumn : indexColumns) {
+                String fqn = indexColumn.name.fqn();
+                for (var input : indexColumn.inputs) {
+                    Object value = input.value();
+                    if (value == null) {
+                        continue;
+                    }
+                    if (value instanceof Iterable<?> it) {
+                        for (Object val : it) {
+                            if (val == null) {
+                                continue;
+                            }
+                            Field field = new Field(fqn, val.toString(), indexColumn.fieldType);
+                            doc.add(field);
                         }
-                        Field field = new Field(fqn, val.toString(), indexColumn.fieldType);
+                    } else {
+                        Field field = new Field(fqn, value.toString(), indexColumn.fieldType);
                         doc.add(field);
                     }
-                } else {
-                    Field field = new Field(fqn, value.toString(), indexColumn.fieldType);
-                    doc.add(field);
                 }
             }
-        }
 
-        for (var constraint : tableConstraints) {
-            constraint.verify(item.insertValues());
+            for (var constraint : tableConstraints) {
+                constraint.verify(item.insertValues());
+            }
+
+            NumericDocValuesField version = new NumericDocValuesField(DocSysColumns.Names.VERSION, -1L);
+            doc.add(version);
+
+            BytesReference source = BytesReference.bytes(xContentBuilder);
+            BytesRef sourceRef = source.toBytesRef();
+            doc.add(new StoredField("_source", sourceRef.bytes, sourceRef.offset, sourceRef.length));
+
+            BytesRef idBytes = Uid.encodeId(item.id());
+            doc.add(new Field(DocSysColumns.Names.ID, idBytes, IdFieldMapper.Defaults.FIELD_TYPE));
+
+            SequenceIDFields seqID = SequenceIDFields.emptySeqID();
+            // Actual values are set via ParsedDocument.updateSeqID
+            doc.add(seqID.seqNo);
+            doc.add(seqID.seqNoDocValue);
+            doc.add(seqID.primaryTerm);
+            return new ParsedDocument(
+                version,
+                seqID,
+                item.id(),
+                doc,
+                source,
+                null,
+                newColumns
+            );
+        } catch (Exception ex) {
+            // Constraints check might throw an exception.
+            // We ensured that XContentBuilder object is closed (in obedience with docs)
+            // by wrapping it into try, can re-throw the exception now.
+            // In case of happy path it will be closed twice, first by BytesReference.bytes(xContentBuilder) and
+            // then close by the try block which will be no-op (since underlying JsonGenerator checks the state).
+            throw ex;
         }
```

Comment:
```suggestion
        }
```

I have the impression that the comment is confusing more than helping

Reply:
throw ex is gone - I think we still need it, typo? 

I restored it.